cli/azd/cmd/up.go

```diff
@@ -111,11 +107,6 @@ func (u *upAction) Run(ctx context.Context) (*actions.ActionResult, error) {
 			output.WithWarningFormat("WARNING: The '--service' flag is deprecated and will be removed in a future release."))
 	}
 
-	err := u.provisioningManager.Initialize(ctx, u.projectConfig.Path, u.projectConfig.Infra)
-	if err != nil {
-		return nil, err
-	}
-
```

Comment:
Part of the provisioning provider initialization process prompts the user for any missing requirements such as subscription id, location or other information required by the provisioning provider.

I believe we opted to initially include this here so that we collected required data as early as possible in the up flow vs waiting for the child provision command to be called.

This likely will work but unclear if this is the user experience we want for this scenario when an environment requirements have not been fully initialized.

Reply:
💯 💯 That's why I said that I'm not fully aware of the way that it was intended to preform.
Let me know what you decide and I will work from there.

Reply:
I'll think on this a bit and let you know what direction to take.

Reply:
> I believe we opted to initially include this here so that we collected required data as early as possible in the up flow vs waiting for the child provision command to be called.

Your belief here is correct, @wbreza.  The goal was to prompt for the subscription and location (and possibly resource group) information before the `package` subtask of `up` started.  Without this code, we would defer these questions until `provision` later on in the `up` process.

Reply:
@wbreza , @ellismg , These are my recommendations for fixing this issue:

- bicep-memory-cache:  If we make the `bicep tool` to cache the `BuildResult` in memory, the bicep-provider can call `bicep.build()` multiple times, but azd would only compile the bicep once.  The `bicep tool` is a singleton within the azd-container. The cache would be only per-azd-invocation.
The only problem here would be if someone wants to use a `hook` in the middle of azd-up (start) and azd-provision and change the infrastructure.  But I'd call that an antipattern :)

- make provisioning provider a singleton:  The providers are currently transient, so we can't keep state within the provider between azd up (start) and azd provision. The side-effect from this change is that we won't be able to support `parallel provisioning` by just injecting the bicep-provider from the container.  We would need to implement a new component that handles multiple provisioning provider instances.  However, it is not too bad, as we are currently not using/supporting parallel deployments.

@john0isaac , What I am sure is that removing the provider from azd up (start) is not the best approach here, as we always push for having all the interactions as soon as possible, and this change would go against that approach.  Let's please keep waiting on @wbreza  and @ellismg to decide how they want to address this.  Thank you for your patience.

Reply:
I believe the path forward here is to make the required changes to make the `provisioning.Manager` a singleton and remove all transient state storage from the manager instance.

One way to achieve this could be the following:
1. Remove `provider`, `options` & `projectPath` from the `provisioning.Manager`
2. Modify `Manager.Initialize` to return the created provider instance `(provisioning.Provider, error)`
3. Update the `Manager` methods in accept a `provisioning.Provider` parameter and the command would act on this instance.

The above gets us back to having a singleton that can be used.

The next step is how can the underlying infra provider be leveraged.  Another idea is to treat the `Manager.Initialize` method as a factory creation method.  Maybe we even rename `Manager.Initialize` to `Manager.CreateProvider`.

Then the manager could store a reference/pointer to the created providers based on the input options of projectPath & infra options.  Any subsequent calls to `Manager.CreateProvider` would return the same transient instance from the first call.  It would still return new instances if params don't match.

Since the providers themselves are transient they can store whatever internal state is required.  So if the bicep build already happened then we can short circuit return out of this.

@ellismg / @vhvb1989 thoughts?